File: javascript/leetcode/450. Delete Node in a BST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */
var deleteNode = function(root, key) {
  if (!root) return root;

  if (key > root.val) root.right = deleteNode(root.right, key);
  else if (key < root.val) root.left = deleteNode(root.left, key);
  else {
    if (!root.left && !root.right) root = null;
    else if (root.left && root.right) {
      const min = small(root.right);
      root.val = min.val;
      root.right = deleteNode(root.right, root.val);
    } else {
      if (!root.right) return root.left;
      else return root.right;
    }
  }

  return root;
};

function small(root) {
  while (root.left) root = root.left;
  return root;
}

File: go/leetcode/450. Delete Node in a BST.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func deleteNode(root *TreeNode, key int) *TreeNode {
    if root == nil{return nil}
    
    if key > root.Val {
        root.Right = deleteNode(root.Right, key)
    } else if key < root.Val {
        root.Left = deleteNode(root.Left, key)
    } else {
        if root.Left == nil && root.Right == nil {
            root = nil
        } else if root.Left != nil && root.Right != nil {
            min := small(root.Right)
            root.Val = min.Val
            root.Right = deleteNode(root.Right, root.Val)
        } else {
            if root.Right == nil {
                root = root.Left
            } else {
                root = root.Right
            }
        }
    }
    
    return root
}

func small(root *TreeNode) *TreeNode {
    for root.Left != nil {root = root. Left}
    return root
}

